lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -1095,6 +1100,54 @@ unsafeEpochNo epochNo
     maxEpochNo :: Word32
     maxEpochNo = fromIntegral @Word31 $ unEpochNo maxBound
 
+-- | Calculate the time at which an epoch begins.
+epochStartTime :: SlotParameters -> EpochNo -> UTCTime
+epochStartTime sps e = slotStartTime sps $ SlotId e 0
+
+-- | Return the epoch immediately before the given epoch, or 'Nothing' if there
+--   is no representable epoch before the given epoch.
+epochPred :: EpochNo -> Maybe EpochNo
+epochPred (EpochNo e)
+    | e == minBound = Nothing
+    | otherwise = Just $ EpochNo $ pred e
+
+-- | Return the epoch immediately after the given epoch, or 'Nothing' if there
+--   is no representable epoch after the given epoch.
+epochSucc :: EpochNo -> Maybe EpochNo
+epochSucc (EpochNo e)
+    | e == maxBound = Nothing
+    | otherwise = Just $ EpochNo $ succ e
+
+-- | For the given time 't', calculate the number of the earliest epoch with
+--   start time 's' such that 't ≤ s'.
+--
+-- Returns 'Nothing' if the calculation would result in an epoch number that is
+-- not representable.
+epochCeiling :: SlotParameters -> UTCTime -> Maybe EpochNo
+epochCeiling sps t
```

Comment:
**[question]**

If we have t=epochStartTime then we endup with the same epoch. I bet it is consistent with how slotCeiling behaves, right? If yes, then I buy this convention;-)

Reply:
> **[question]**
> 
> If we have t=epochStartTime then we endup with the same epoch. I bet it is consistent with how slotCeiling behaves, right? If yes, then I buy this convention;-)

Yes, this is consistent with how `slotCeiling` behaves!

It's also consistent with the way that `ceiling` works in general:

```hs
> import Prelude
> ceiling (1.0 :: Double) :: Integer
1
> ceiling (1.00000001 :: Double) :: Integer
2
```